daemon/network.go

```diff
@@ -99,15 +100,45 @@ func (daemon *Daemon) getAllNetworks() []libnetwork.Network {
 	return daemon.netController.Networks()
 }
 
-func isIngressNetwork(name string) bool {
-	return name == "ingress"
+type ingressJob struct {
+	create *clustertypes.NetworkCreateRequest
+	ip     net.IP
 }
 
-var ingressChan = make(chan struct{}, 1)
+var (
+	ingressWorkerOnce  sync.Once
+	ingressJobsChannel chan *ingressJob
+	ingressWorkerStop  chan struct{}
+	ingressID          string
+)
+
+func (daemon *Daemon) setupIngressWorker() {
+	ingressWorkerStop = make(chan struct{}, 1)
+	ingressJobsChannel = make(chan *ingressJob, 100)
+	go func() {
+		for {
+			select {
+			case <-ingressWorkerStop:
+				return
+			case r := <-ingressJobsChannel:
+				if r.create != nil {
+					daemon.setupIngress(r.create, r.ip, ingressID)
+					ingressID = r.create.ID
+				} else {
+					daemon.releaseIngress(ingressID)
+					ingressID = ""
+				}
+			}
+		}
+	}()
+}
 
-func ingressWait() func() {
-	ingressChan <- struct{}{}
-	return func() { <-ingressChan }
+func (daemon *Daemon) stopIngressWorker() {
+	if ingressWorkerStop != nil {
+		close(ingressWorkerStop)
+		ingressWorkerOnce = sync.Once{}
```

Comment:
> Is there an external guarantee that stopIngressWorker can't be called concurrently with SetupIngress or ReleaseIngress?

Yes, `stopIngressWorker` is called when docker leaves the swarm.
`SetupIngress` and `ReleaseIngress` cannot be called concurrently as they are all queued.
The worker logic is in fact to serialize the setup/release events.

I was under the impression setup or release which in turn are called by executor.Configure() cannot happen to be called after `DaemonLeavesCluster()`.

Given the worker drains ingress network creation/deletion events, I did not expect a long list of events left in the queue to process.